lineapy/api/artifact_serializer.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import sys
+import types
+from inspect import getmodule
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+from pandas.io.pickle import to_pickle
+
+from lineapy.data.types import ARTIFACT_STORAGE_BACKEND, LineaID
+from lineapy.exceptions.db_exceptions import ArtifactSaveException
+from lineapy.plugins.mlflow_io import mlflow_io
+from lineapy.plugins.utils import slugify
+from lineapy.utils.analytics.event_schemas import ErrorType, ExceptionEvent
+from lineapy.utils.analytics.usage_tracking import track
+from lineapy.utils.config import options
+from lineapy.utils.logging_config import configure_logging
+
+try:
+    import mlflow
+except ImportError:
+    pass
+
+
+logger = logging.getLogger(__name__)
+configure_logging()
+
+
+def serialize_artifact(
+    value_node_id: LineaID,
+    execution_id: LineaID,
+    reference: Any,
+    name: str,
+    storage_backend: Optional[ARTIFACT_STORAGE_BACKEND] = None,
+    **kwargs,
+) -> Dict[str, Any]:
+    """
+    Serialize artifact using various backend.
+
+    Currently, most objects are using lineapy as the backend for serialization.
+    The only exception is mlflow supported model flavors. In order to use
+    mlflow for ml model serialization, following conditions need to be
+    satisified:
+    1. the artifact(ML model) should be a mlflow supported flavor
+    2. mlflow is installed
+    3. storage_backend should be mlflow or storage_backend is None and
+    `options.get("default_ARTIFACT_STORAGE_BACKEND")=='mlflow'`
+
+    Parameters
+    ----------
+    value_node_id: LineaID
+        Value node id in Linea Graph
+    execution_id: LineaID
+        Execution id
+    reference: Union[object, ExternalState]
+        Same as reference in :func:`lineapy.api.save`
+    name: str
+        Same as reference in :func:`lineapy.api.save`
+    storage_backend: Optional[ARTIFACT_STORAGE_BACKEND]
+        Same as reference in :func:`lineapy.api.save`
+    **kwargs:
+        Same as reference in :func:`lineapy.api.save`
+
+    Returns
+    -------
+    Dict
+        returned a dictionary with following key-value pair
+        backend: storage backend used to save the artifact
+        metadata: metadata of the storage backed
+    """
+    if options.get("mlflow_tracking_uri") is not None:
+        if storage_backend == ARTIFACT_STORAGE_BACKEND.mlflow or (
+            storage_backend is None
+            and options.get("default_ml_models_storage_backend")
+            == ARTIFACT_STORAGE_BACKEND.mlflow
+        ):
+            if "mlflow" not in sys.modules:
```

Comment:
Can we break out this MLflow specific logic to a helper function the way `_try_write_to_pickle` is? 

Ideally this function  `serialize_artifact` will just check to see which serializer to use and call the corresponding helper function, which will keep it lightweight and re-usable.

Reply:
Sure, thought about this during implementation.